python/neuroglancer/write_annotations.py

```diff
@@ -197,20 +407,34 @@ def _add_obj(self, coords: Sequence[float], id: Optional[int], **kwargs):
             id=id, encoded=encoded.tobytes(), relationships=related_ids
         )
 
+        for i in range(int(n_spatial_coords)):
+            chunk_index = self.get_chunk_index(
```

Comment:
For annotation types other than a point, we need to add all of the intersecting chunks.  For example, for a line we need all of the chunks that intersect the line, not just the endpoints, as otherwise when viewing some portion of the middle of the line it might not be displayed in Neuroglancer.

There are probably various approaches that could be taken to compute the set of intersecting chunks, but the approach I took was to compute the bounding box in grid space, and then for each chunk within the bounding box do the exact line or ellipsoid intersection test.  For axis_aligned_bounding_box no exact test is needed since the bounding box is exact.

Reply:
How is this handled during downsampling? Each annotation can appear multiple times per downsampling level, but should appear in all chunked it overlaps with within that level? So does one randomly choose annotations amongst the entire level to show at that level or does one choose randomly amongst the anntotations that intersect with the chunk?  If the latter how do you resolve that with the requirement that it be listed for all chunks that it intersects with?  Does one first do the sampling by chunks, but then take that filtered list and do the spatial overlap test to see if it gets listed in any spatial chunk? Therefore sometimes ending up with more annotations in a spatial chunk than the maximum that you were going for?

Reply:
This is a description of how the spatial index is computed:

```
// Writes the multi-level spatial index.  Returns the spatial index metadata.
//
// This involves the following steps:
//
// 1. To initialize the `remaining_annotations` table, key each input annotation
//    with the all-zero vector as its grid cell.
//
// 2. For each spatial index level `level_i`, starting at the coarsest level
//    (`level_i=0`):
//
//    2.1. For each dimension with variable chunk size: if `level_i == 0`, the
//         grid cell size is equal to the full range of the dimension; if
//         `level_i != 0`, the grid cell size is equal to either 1 or 1/2 times
//         the cell size for `level_i - 1`, whichever leads to a more isotropic
//         chunk (the dimension with the largest cell size is always halved).
//
//    2.2. Compute the `keyed_by_cell` table containing for each annotation in
//         `remaining_annotations` keyed by its `parent_cell` in level
//         `level_i - 1` (or the root all-zero vector grid cell if
//         `level_i == 0`) a copy of the annotation keyed by each child grid
//         cell in `level_i` it intersects.
//
//    2.3. Compute `max_per_cell`, the maximum number of annotations per cell.
//
//    2.4. Compute the `output_by_cell` table containing a random
//         `min(1, max_annotations_per_spatial_cell / double(max_per_cell))`
//         fraction of `keyed_by_cell`.  Also compute the updated value of
//         `remaining_annotations`, equal to the elements of `keyed_by_cell` not
//         included in `output_by_cell`.
//
//    2.5. Write the contents of `output_by_cell` as the `level_i` spatial
//         index.
```

The sampling is done based on the annotation id, using a hash, which ensures that an annotation is picked in all cells if it is picked in one cell.

We can't exactly limit the number of annotations per cell but it is a binomial distribution and assuming the hash is reasonable won't be too high.

We have to pick a single sampling probability for all cells to ensure that the relative density of different regions is preserved.

Annotations that span a large number of cells are problematic, since then adding a spatial index level does not necessarily reduce the number of annotations per cell.  Hopefully you don't need that case to work.  A different type of spatial index would probably needed in that case.

Reply:
I've now added a general purpose nd index (rtree) to the code to keep track of where annotations are so they can be written into the appropriate chunks.

 I've added lower_bound and upper_bound parameters to the _add_obj and _add_two_point_obj functions, with custom values set depending on the annotation type.  
 
 this should set the code up better for doing downsampling as this general spatial index can be used across the scales to query for annotations that exist in different spatial boxes.